rse-py-scripting.Rmd

```diff
@@ -103,42 +133,461 @@ optional arguments:
+you're probably doing something wrong [@Xu2015]:
+most users never use most configuration options and find their presence confusing.
+
+## Where should configuration files go? {#rse-py-scripting-config-loc}
+
+System-wide settings for an application called `app` are often stored in `/etc/app.yml`.
+(Older programs sometimes use a name like `app.rc` for their configuration file,
+where `rc` stands for "resource control".)
+Alternatively,
+some programs will set an environment variable (Section \@ref(rse-bash-advanced-vars))
+to the name of the installation directory,
+and then read the system configuration file from there.
+For example,
+if the application's environment variable is `THESIS`,
+we can read configuration from:
+
+```python
+install_dir = os.getenv('THESIS')
+config_file = os.path.join(install_dir, 'config.yml')
+if os.path.isfile(config_file):
+    with open(config_file, 'r') as reader:
+        config = yaml.load(reader)
 ```
 
-## Where should I put my code? {#rse-py-scripting-code}
+Similarly,
+we can get personal settings from `$HOME/.thesis.yml`;
+the leading '.' hides the configuration file from `ls`.
+Finally,
+per-job settings can come from `thesis.yml` in the current directory,
+where again "app" is replaced with the name of the program.
+
+## How can I keep a record of the actual configuration that produced particular results? {#s:configure-dump}
+
+We don't need to write any of this when we first create a new tool.
+However,
+experience teaches that little tools grow into big ones.
+As soon as we have more than a couple of configuration parameters,
+or as soon as some of them are usually set to the same value,
+we should start to think about saving our settings in files.
+
+This is particularly important if we want others (including our future self) to be able to reproduce our work.
+Careful record keeping is essential to reproducible science,
+and if we add just a few lines to our program,
+the computer can save the entire configuration for a particular run of a program.
+For example,
+if we have stored our settings in a Python dictionary called `settings` like this:
 
-Using this template as a starting point,
-we can add the functions we developed previously to a script called...
+```python
+settings = {
+    'logfile': '/tmp/thesis.log',
+    'quiet': False,
+    'overwrite': True,
+    'fonts': ['Comic Sans']
+}
+```
+
+we can save that as YAML using this short function:
+
+```python
+def save_config(filename, settings):
+    '''Save configuration settings to a file.'''
+
+    with open(filename, 'w') as writer:
+        yaml.dump(settings, writer)
+
+save_config('thesis.yml', settings)
+```
+
+```shell
+$ cat thesis.yml
+```
 
-FIXME: In this section we will take the code developed in the novice book
-and turn it into a Python script that can be executed at the command line.
-See [this lesson](https://carpentrieslab.github.io/python-aos-lesson/04-cmdline/index.html)
-for an example of this approach in action.
```

Comment:
Once we've settled #240 (i.e. the narrative that runs through the RSE book), I'd argue that this placeholder about taking the code we've developed (as part of the narrative) and turning it into a command line program is still relevant.

Reply:
:+1: